File: packages/recite-core/src/recite/core/text.py
"""Editing documents by span, without ever corrupting one.

Corrections arrive from many independent rules and refer to offsets in the
*original* text. Applying them naively in document order would invalidate every
offset after the first edit, and two rules that both want to rewrite the same
citation would silently produce nonsense.

:func:`apply_corrections` therefore applies edits back-to-front and refuses any
correction that overlaps one it already accepted, reporting the rejects instead
of dropping them.
"""

from __future__ import annotations

import difflib
from dataclasses import dataclass, field

from .models import Correction, Span

__all__ = ["PatchResult", "apply_corrections", "line_col", "unified_diff"]


@dataclass
class PatchResult:
    """Outcome of applying a batch of corrections to one document."""

    text: str
    """The patched document."""

    applied: list[Correction] = field(default_factory=list)
    skipped: list[tuple[Correction, str]] = field(default_factory=list)
    """``(correction, reason)`` for each edit that was not applied."""

    @property
    def changed(self) -> bool:
        return bool(self.applied)


def apply_corrections(text: str, corrections: list[Correction]) -> PatchResult:
    """Apply non-overlapping corrections to ``text``.

    Corrections are considered in document order so that when two of them
    collide, the earlier one wins — that keeps the result stable regardless of
    which rule happened to run first. No-ops (a replacement identical to what
    is already there) are dropped quietly, since a rule proposing one is not an
    error, just redundant.
    """
    ordered = sorted(corrections, key=lambda c: (c.span.start, c.span.end))

    accepted: list[Correction] = []
    skipped: list[tuple[Correction, str]] = []

    for correction in ordered:
        if correction.span.end > len(text):
            skipped.append((correction, "span extends past end of document"))
            continue
        if correction.is_noop(text):
            skipped.append((correction, "replacement is identical to source"))
            continue
        clash = next(
            (a for a in accepted if a.span.overlaps(correction.span)),
            None,
        )
        if clash is not None:
            skipped.append(
                (correction, f"overlaps an earlier fix at {clash.span.as_tuple()}")
            )
            continue
        accepted.append(correction)

    patched = text
    for correction in sorted(accepted, key=lambda c: c.span.start, reverse=True):
        patched = (
            patched[: correction.span.start]
            + correction.replacement
            + patched[correction.span.end :]
        )

    return PatchResult(text=patched, applied=accepted, skipped=skipped)


def line_col(text: str, offset: int) -> tuple[int, int]:
    """1-based ``(line, column)`` for a character offset.

    Used for human-readable output and for SARIF, both of which count from 1.
    """
    if offset < 0:
        raise ValueError("offset must be non-negative")
    offset = min(offset, len(text))
    line = text.count("\n", 0, offset) + 1
    line_start = text.rfind("\n", 0, offset) + 1
    return line, offset - line_start + 1


def unified_diff(before: str, after: str, path: str = "document") -> str:
    """A standard unified diff, empty when nothing changed."""
    if before == after:
        return ""
    diff = difflib.unified_diff(
        before.splitlines(keepends=True),
        after.splitlines(keepends=True),
        fromfile=f"a/{path}",
        tofile=f"b/{path}",
        n=2,
    )
    return "".join(diff)


def snippet(text: str, span: Span, width: int = 60) -> str:
    """One line of context around ``span``, for terminal output."""
    start = max(0, span.start - width // 2)
    end = min(len(text), span.end + width // 2)
    fragment = text[start:end].replace("\n", " ").strip()
    prefix = "…" if start > 0 else ""
    suffix = "…" if end < len(text) else ""
    return f"{prefix}{fragment}{suffix}"

File: packages/core/src/text.ts
/**
 * Editing documents by span, without ever corrupting one.
 *
 * Corrections arrive from independent rules and refer to offsets in the
 * *original* text. Applying them in document order would invalidate every
 * offset after the first edit, and two rules that both want to rewrite the
 * same citation would silently produce nonsense.
 *
 * {@link applyCorrections} therefore applies edits back-to-front and refuses
 * any correction that overlaps one it already accepted, reporting the rejects
 * instead of dropping them.
 */

import type { Correction, Span } from "./model.js";
import { sliceSpan, spansOverlap } from "./model.js";

export interface PatchResult {
  readonly text: string;
  readonly applied: readonly Correction[];
  /** `[correction, reason]` for every edit that was not applied. */
  readonly skipped: ReadonlyArray<readonly [Correction, string]>;
  readonly changed: boolean;
}

export function applyCorrections(
  text: string,
  corrections: readonly Correction[],
): PatchResult {
  const ordered = [...corrections].sort(
    (a, b) => a.span.start - b.span.start || a.span.end - b.span.end,
  );

  const applied: Correction[] = [];
  const skipped: Array<readonly [Correction, string]> = [];

  for (const correction of ordered) {
    if (correction.span.end > text.length) {
      skipped.push([correction, "span extends past end of document"]);
      continue;
    }
    if (sliceSpan(text, correction.span) === correction.replacement) {
      skipped.push([correction, "replacement is identical to source"]);
      continue;
    }
    // Decide in document order so the outcome never depends on which rule
    // happened to run first.
    const clash = applied.find((a) => spansOverlap(a.span, correction.span));
    if (clash) {
      skipped.push([
        correction,
        `overlaps an earlier fix at [${clash.span.start}, ${clash.span.end})`,
      ]);
      continue;
    }
    applied.push(correction);
  }

  let patched = text;
  for (const correction of [...applied].sort((a, b) => b.span.start - a.span.start)) {
    patched =
      patched.slice(0, correction.span.start) +
      correction.replacement +
      patched.slice(correction.span.end);
  }

  return { text: patched, applied, skipped, changed: applied.length > 0 };
}

/** One line of context around a span, for display. */
export function snippet(text: string, s: Span, width = 72): string {
  const start = Math.max(0, s.start - Math.floor(width / 2));
  const end = Math.min(text.length, s.end + Math.floor(width / 2));
  const fragment = text.slice(start, end).replace(/\s+/g, " ").trim();
  return `${start > 0 ? "…" : ""}${fragment}${end < text.length ? "…" : ""}`;
}

/**
 * A minimal line-level diff, good enough to show what a fix changed.
 *
 * Deliberately not a full Myers diff: the output is read by a human comparing
 * a handful of edited lines, and pulling in a diff library for that would be
 * more dependency than the job needs.
 */
export function lineDiff(before: string, after: string): string {
  if (before === after) return "";

  const a = before.split("\n");
  const b = after.split("\n");
  const out: string[] = [];

  const max = Math.max(a.length, b.length);
  for (let i = 0; i < max; i++) {
    const left = a[i];
    const right = b[i];
    if (left === right) continue;
    if (left !== undefined) out.push(`- ${left}`);
    if (right !== undefined) out.push(`+ ${right}`);
  }

  return out.join("\n");
}
